server/server.go

```diff
@@ -65,59 +227,180 @@ func readFormData(r *http.Request) string {
-				log.Println(err)
-				return responseString
-			}
-			fmt.Println(header.Name)
-		}
-	} else {
-		// For zip file.
+
+	// Read the file based on the type of file compression.
+	assignTestingInfo.RootDir = strings.TrimSuffix(handler.Filename, path.Ext(handler.Filename))
+
+	if http.DetectContentType(fileHeader) == constants.ZipMimeFileType {
+		// Read zip file.
 		unZipped, err := zip.NewReader(file, handler.Size)
 		if err != nil {
+
 			responseString := `"Unzip Error":"Error in unzipping uploaded file"`
-			log.Println(err)
+			log.Println("error in unzipping file", err)
 			return responseString
 		}
-		for _, file := range unZipped.File {
-			fmt.Println(file.Name)
+		return storeUnzippedFiles(unZipped)
+
+	} else if http.DetectContentType(fileHeader) == constants.TarGzMimeFileType {
+		// Read tar.gz file.
+		assignTestingInfo.RootDir = strings.TrimSuffix(assignTestingInfo.RootDir,
+			path.Ext(assignTestingInfo.RootDir))
+		fileReader, err := handler.Open()
+		gZipReader, err := gzip.NewReader(fileReader)
+		if err != nil {
+			responseString := `"Untar Error":"Error in untaring uploaded file"`
+			log.Println("error in untaring file", err)
+			return responseString
 		}
+		unTarred := tar.NewReader(gZipReader)
+		return storeUnTarredFiles(unTarred)
+
+	} else {
+		// Read tar file.
+		var fileReader io.ReadCloser = file
+		unTarred := tar.NewReader(fileReader)
+		return storeUnTarredFiles(unTarred)
 	}
+}
 
+// storeUnTarredFiles stores unTared files to 'assignments' directory.
+func storeUnTarredFiles(unTarred *tar.Reader) string {
+
+	errResponse := `"UnTar Error":"Error in un-tarring uploaded file"`
+	dest := filepath.Join(constants.AssignmentsDir, assignTestingInfo.RootDir)
+	for {
+		header, err := unTarred.Next()
+		if err == io.EOF {
+			// End of tar file.
+			break
+		}
+		if err != nil {
+			log.Println("unTar error: ", err)
+			return errResponse
+		}
+
+		filename := header.Name
+		switch header.Typeflag {
+		case tar.TypeDir:
+			err := os.MkdirAll(filepath.Join(dest, filename), os.FileMode(header.Mode)) // or use 0755 if you prefer
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+		case tar.TypeReg:
+			err := os.MkdirAll(filepath.Join(dest, filepath.Dir(filename)), os.FileMode(header.Mode))
+			writer, err := os.Create(filepath.Join(dest, filename))
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+			_, err = io.Copy(writer, unTarred)
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+			err = os.Chmod(filepath.Join(dest, filename), os.FileMode(header.Mode))
+
+			if err != nil {
+				log.Println("unTar error: ", err)
+				return errResponse
+			}
+
+			writer.Close()
+		default:
+			log.Println("unable to unTar type : ", header.Typeflag)
+			return errResponse
+		}
+	}
 	return ""
 }
 
+// storeUnzippedFiles stores unzipped files to 'assignments' directory.
```

Comment:
inaccurate comment. Refer to comment on `storeUntarredFiles`

Reply:
Fixed